CI: MVD CI/CD Setup Using Docker. Without cloud dependencies (#201)
agera-edc/MinimumViableDataspace: system-tests/README.md

## System tests

The test copy a file from provider to consumer blob storage account.

### Building MVD project

MVD dependencies are Eclipse DataSpaceConnector(EDC) and Registration Service. Both of these dependencies are not published to any central artifactory yet so in local
development we have to use locally published dependencies, once this is done MVD can be build using

```bash

> ./gradlew build

Reviewer: As discussed - it would make sense to exclude the test task here, otherwise the CatalogClientTest fails

Author: Updated. Thank you.

```

#### Publish EDC and Registration Service to local Maven

Checkout [Eclipse DataSpaceConnector repository](https://github.com/eclipse-dataspaceconnector/DataSpaceConnector).

Publish EDC libraries to local Maven artifactory by executing gradle command `./gradlew publishToMavenLocal -Pskip.signing` from EDC root
folder.

Checkout [Registration Service repository](https://github.com/eclipse-dataspaceconnector/RegistrationService).

Publish Registration Service libraries to local Maven artifactory by executing gradle command `./gradlew publishToMavenLocal` from Registration Service root
folder.

### Running test locally

MVD System tests can be executed locally against a local MVD instance. MVD runs three EDC Connectors and one Registration Service.

First please make sure that you are able to build MVD locally as described in [Building MVD project](#building-mvd-project) section.

- We need to build EDC Connector launcher and Registration Service launcher.

> - Go to EDC root folder. And execute

Reviewer: Going to EDC root folder didn't work with error `Task 'shadowJar' not found in project ':launchers'.`.

Should we go to MVD root folder?

Author: Yes, updated. Thanks.

Good catch.

    ```bash
    ./gradlew -DuseFsVault="true" :launcher:shadowJar
    ```

- Go to Registration service root folder. And execute

    ```bash
    ./gradlew :launcher:shadowJar
    ```

- Update Registration service launcher path in `system-tests/docker-compose.yml` file. Look for section `#UPDATE_REGISTRATION_SERVICE_LAUNCHER_PATH_HERE#` and update it e.g. `/home/user/RegistrationService/launcher`.

- Start MVD using docker-compose.yml file.

    ```bash
    docker-compose -f system-tests/docker-compose.yml up --build
    ```

- This will start three EDC Connectors, one Registration Service, one HTTP Nginx Server to serve DIDs, Azurite blob storage service and also will seed initial required data using a [postman collection](../deployment/data/MVD.postman_collection.json).

- `newman` docker container will automatically stop after seeding initial data from postman scripts.

> - EDC Connectors needs to be registered using Registration Service CLI client jar. This client jar must be available under `RegistrationService-Root/client-cli/build/libs` folder.

Reviewer: ```suggestion
- EDC Connectors needs to be registered using Registration Service CLI client jar. After publishing RegistrationService locally the client jar should be available under `RegistrationService-Root/client-cli/build/libs` folder.
```
nit: suggestion to make it clear that if the Reg Service is published locally then there's no new action needed, the client jar should be there

> - EDC Connectors needs to be registered using Registration Service CLI client jar. After publishing RegistrationService locally the client jar should be available under `RegistrationService-Root/client-cli/build/libs` folder.

Reviewer 2: The step-by-step instructions could be replaced by small scripts that automate most of it. Have you considered this? I think we could have:
- `runMVD.local.sh`: builds dependencies and starts MVD
-  `runTests.local.sh`: registers participants and runs tests

Author: @marcgs Yes could be. But how we will maintain & validate such script ? Because we are not using it in CI. Only script used here is to register participant and that is covered as it is also used in CI.

(Or if you want, we can discuss this one more when we do the next feature to run MVD locally including UI, as this PR focuses more on CI/CD topic.)

Reviewer 2: I would use the same scripts to run the tests against local MVD on CI. IMO the main advantage would be to get rid of the many steps for running MVD locally which can lead to errors, but it is not a must have either.

Author: As discussed offline leaving it as it as in this PR. As it can be addressed later. Thanks

    ```bash
    export REGISTRATION_SERVICE_CLI_JAR_PATH=registration service client jar path

>     system-tests/resources/register-participants.sh

Reviewer 2: On linux and Mac you need either`./` at the front or `bash XXX`

Author: Updated. Thanks
